Library/Homebrew/extend/os/linux/requirements/java_requirement.rb

```diff
@@ -0,0 +1,31 @@
+require "language/java"
+
+class JavaRequirement < Requirement
+  default_formula "jdk"
+
+  env do
+    next unless @java_home
+    env_java
+    java_home = Pathname.new(@java_home)
+    if (java_home/"include").exist? # Oracle JVM
+      env_oracle("linux")
+    end
+  end
+
+  private
+
+  def preferred_java
```

Comment:
> Is it worth preferring this over JAVA_HOME?

Yes. It would solve a potential practical problem:

1. Someone is running some proprietary app that's only supported on Java 7 and has to set JAVA_HOME to enforce this.
2. They try to install `gosu`, which requires 8+
3. They install `jdk` to meet the requirement
4. But since `JAVA_HOME` is preferred, they **still** don't meet the requirement and are confused/frustrated.

> I think regardless this logic could perhaps be shared between platforms as Homebrew won't have a jdk formula

Works for me. With that in mind, how about the following order?

1. Use the `jdk` formula (with a comment explaining it's provided by Linuxbrew)
2. Use `java_home_cmd` to guess. This would return `nil` in the generic code and have a macOS-specific override (which I've already written)
3. Use `JAVA_HOME`
4. Use `which "java"`